### PR TITLE
[WIP] Add LoRA as efficient finetuning strategy

### PR DESCRIPTION
**Pull request by Raldir**
Added the option to apply LoRA (https://arxiv.org/pdf/2106.09685.pdf) to Huggingface text model as an additional option for efficient fine-tuning, see PR #1666.  

- Added flag for model: `model.hf_text.adaptation=lora`. This currently only supports loRA but can be extended to adapters or similar adaptations to transformers that modify the architecture directly. LoRA is currently set to be inserted only into Query and Value attention weights of a transformer, as done in their paper and shown to be most effective.
-  While loRA is currently only added to hf_text, should be extendable to most other transformer architecture, as soon as the general structure stands. As the adaptation is currently coded into the model class directly, might consider adding a separate adaptation class to be then added to different models.
- For optimizer `optimization.efficient_finetuned` added flag `lora` and `lora_bias`. The first fine-tunes only loRA weights while the second updates both loRA weights and all biases of the transformer. The latter has not been comprehensively explored in the original paper but has been suggested as an additional modification.
- LoRA uses two hyperparameters. Currently, these are set also through the model config. However, this is rather a hack. Since adapters etc. also use hyperparameters, maybe consider adding a separate config file for adaptations?   
- Currently, when saving the model the entire model is stored, which is unnecessary. TODO: Need to change to solely save fine-tuned weights, also applies to bit_fit and norm_fit.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by Raldir** [comment]
Just to say that experiments on how LoRA compares to bit_fit and norm_fit are ongoing.

File: text/src/autogluon/text/automm/models/lora_layers.py [hunk 1]
```diff
@@ -0,0 +1,373 @@
+
```
Comment (sxjscience) [inline text/src/autogluon/text/automm/models/lora_layers.py:1]: Mention that this file is modified based on https://github.com/microsoft/LoRA/blob/main/loralib/layers.py

File: text/src/autogluon/text/automm/models/lora_layers.py [hunk 2]
```diff
@@ -0,0 +1,373 @@
...
+import math
+from typing import List
+
+
+def identity(x):
+    return x
+
+
+class LoRALayer():
```
Comment (sxjscience) [inline text/src/autogluon/text/automm/models/lora_layers.py:19]: Consider to just use the following (remove parenthesis).

```
class LoRALayer:

```

File: text/src/autogluon/text/automm/models/huggingface_text.py [hunk 3]
```diff
@@ -14,12 +14,24 @@
...
 logger = logging.getLogger(AUTOMM)
 
 
+def replace_layers_with_lora(model, old, lora_r, lora_alpha):
+    for n, module in model.named_children():
+        if len(list(module.children())) > 0:
+            replace_layers_with_lora(module, old, lora_r, lora_alpha)
+
+        if old in n:
+            lora_layer = LoRALinear(module.in_features, module.out_features, r= lora_r, lora_alpha=lora_alpha, merge_weights=False)
```
Comment (sxjscience) [inline text/src/autogluon/text/automm/models/huggingface_text.py:29]: We need to detect the type of module. Should we do the following instead? 

```
if isinstance(module, nn.Linear):
   lora_layer = LoRALinear(...)
```

**Comment by sxjscience** [comment]
We recently enabled black to check the coding style so you will have to rebase the code with the latest master.

File: text/src/autogluon/text/automm/models/huggingface_text.py [hunk 4]
```diff
@@ -14,12 +14,24 @@
...
     get_column_features,
 )
-
+from .lora_layers import LoRALinear
 hf_logging.set_verbosity_error()
 
 logger = logging.getLogger(AUTOMM)
 
 
+def replace_layers_with_lora(model, old, lora_r, lora_alpha):
```
[the remark this answers is not in the document]
Comment (zhiqiangdon) [reply text/src/autogluon/text/automm/models/huggingface_text.py:23]: Also, we need to add docstrings and type hints for arguments. Please refer to other functions for examples.

File: text/src/autogluon/text/automm/models/huggingface_text.py (hunk 3 above)
Comment (Raldir) [reply text/src/autogluon/text/automm/models/huggingface_text.py:29]: I agree that relying on the name is brittle. That said, LoRA is not applied to all linear layers, only query and value attention weights are replaced. I was thinking we might be able to maybe make a check on the type of parent module? But again here, different models define their own attention module (e.g. RobertaSelfAttention, DisentangledSelfAttention (Deberta)) so we might have to explicitly state which modules we support LoRA for?

File: text/src/autogluon/text/automm/optimization/utils.py [hunk 5]
```diff
@@ -11,13 +11,14 @@
...
 from ..constants import (
-    BINARY, MULTICLASS, REGRESSION, MAX, MIN, NORM_FIT, BIT_FIT,
+    BINARY, MULTICLASS, REGRESSION, MAX, MIN, NORM_FIT, BIT_FIT, LORA, LORA_BIAS,
     ACC, ACCURACY, RMSE, ROOT_MEAN_SQUARED_ERROR, R2, QUADRATIC_KAPPA,
     ROC_AUC, AVERAGE_PRECISION, LOG_LOSS, CROSS_ENTROPY,
     PEARSONR, SPEARMANR,
 )
 import warnings
 from .soft_target_crossentropy import SoftTargetCrossEntropy
+from ..models.lora_layers import LoRALayer
```
Comment (zhiqiangdon) [inline text/src/autogluon/text/automm/optimization/utils.py:21]: Looks like `LoRALayer` is unused in this file.

File: text/src/autogluon/text/automm/models/huggingface_text.py [hunk 6]
```diff
@@ -53,12 +69,23 @@ def __init__(
...
         self.model = AutoModel.from_pretrained(checkpoint_name)
+        self.adaptation = adaptation
+        self.lora_r = lora_r
+        self.lora_alpha = lora_alpha
+
+        if self.adaptation == 'lora':
+            print('Applying LoRA adaptation to attention layers...')
+            replace_layers_with_lora(self.model, "query", self.lora_r, self.lora_alpha)
+            replace_layers_with_lora(self.model, "value", self.lora_r, self.lora_alpha)
+
```
Comment (zhiqiangdon) [inline text/src/autogluon/text/automm/models/huggingface_text.py:88]: This layer replacement operations can be moved outside of specific model classes. We can move them to the end of `create_model`, i.e., after the model is created. In this way, we can apply lora to any models, not limited to text models.

Comment (Raldir) [reply text/src/autogluon/text/automm/models/huggingface_text.py:88]: Good idea!

File: text/src/autogluon/text/automm/configs/model/fusion_mlp_image_text_tabular.yaml [hunk 7]
```diff
@@ -62,6 +62,9 @@ model:
     insert_sep: True
     text_segment_num: 2
     stochastic_chunk: False
+    adaptation: null # Select from options: Lora
+    lora_r: 8 #TODO: Hacky way of changing LoRA hyperparameter. Seperate config file for all adaptation approaches (e.g.. adapters, LoRA...)?
+    lora_alpha: 8
```
[the remark this answers is not in the document]
Comment (sxjscience) [reply text/src/autogluon/text/automm/configs/model/fusion_mlp_image_text_tabular.yaml:67]: Moving to `optimization` sounds good to me.

File: text/src/autogluon/text/automm/optimization/utils.py [hunk 8]
```diff
@@ -11,13 +11,14 @@
     get_linear_schedule_with_warmup,
 )
 from ..constants import (
-    BINARY, MULTICLASS, REGRESSION, MAX, MIN, NORM_FIT, BIT_FIT,
+    BINARY, MULTICLASS, REGRESSION, MAX, MIN, NORM_FIT, BIT_FIT, LORA, LORA_BIAS,
```
Comment (sxjscience) [inline text/src/autogluon/text/automm/optimization/utils.py:14]: Considering to add LORA_NORM (LoRA + NormFit).

File: .gitignore [hunk 9]
```diff
@@ -123,6 +123,7 @@ train/*
 tutorials/checkpoint/*
 tutorials/train/*
 examples/image_classification/*.csv
+examples/few_shot/*
```
Comment (sxjscience) [inline .gitignore:126]: Remove this line

**Comment by sxjscience** [comment]
The design looks good to me overall. Two remaining items:
1. Add a test-case about LoRA in this function https://github.com/awslabs/autogluon/blob/cc8362684ed4412dabe5a2114cd35a1cfce1c8d3/text/tests/unittests/automm/test_automm_predictor.py#L65-L93 . You may change some of the BIT_FIT or NORM_FIT to LORA_BIAS or LORA_NORM.
2. Resolve conflicts.

File: text/src/autogluon/text/automm/utils.py [hunk 10]
```diff
@@ -703,6 +708,31 @@ def create_model(
...
+        A PyTorch model.
+    config:
+        A DictConfig object. The optimization config should be accessible by "config.optimization".
+    """
+    if 'lora' in config.optimization.efficient_finetune:
+        model = inject_lora_to_linear_layer(
+            model = model,
+            lora_r = config.optimization.lora.r,
+            lora_alpha = config.optimization.lora.alpha,
+            filter = ['query', 'value'] # Fine-tune only Query And Value Attention, recommended in https://arxiv.org/abs/2106.09685
```
Comment (sxjscience) [inline text/src/autogluon/text/automm/utils.py:730]: potentially make `filter` a flag in config.

File: text/src/autogluon/text/automm/models/utils.py [hunk 11]
```diff
@@ -282,39 +283,49 @@ def assign_layer_ids(
...
     )
     return name_to_id, left_names
 
 
 def get_column_features(
-    batch: Dict[str, torch.Tensor],
-    column_name_prefix: str,
-    features: torch.Tensor,
-    valid_lengths: torch.Tensor,
+        batch: Dict[str, torch.Tensor],
```
Comment (sxjscience) [inline text/src/autogluon/text/automm/models/utils.py:325]: It might be related to the black formatting issue (indentation).

**Comment by Raldir** [comment]
Opened PR #1809 for an easier merge with master.

**Pull request closed without merging**